docs/storage/commit_log/commit_log_based_storage_design.md

```diff
@@ -0,0 +1,354 @@
+    glog.Errorf("should not happen - local DB has different data than STHs topic")
+    return
+  }
+
+  // Look to see if anyone else has already stored data just ahead of our STH.
+  nextOffset := dbSTH.stdOffset
+  nextSTH := nil
+  for {
+      nextOffset++
+      nextSTH = kafka.Read("STHs/<treeID>", nextOffset)
+      if nextSTH == nil {
+        break
+      }
+      if nextSTH.expectedOffset != nextOffset {
+        // Someone's been writing STHs when they weren't supposed to be, skip
+        // this one until we find another which is in-sync.
+        glog.Warning("skipping invalid STH")
+        continue
+      }
+      if nextSTH.timestamp < ourSTH.timestamp || nextSTH.tree_size < ourSTH.tree_size {
+        glog.Fatal("should not happen - earlier STH with later offset")
+        return
+    }
+  }
+
+  if nextSTH == nil {
+    // We're up-to-date with the STHs topic (as of a moment ago) ...
+    if !IsMaster() {
+      // ... but we're not allowed to create fresh STHs.
+      return
+    }
+    // ... and we're the master. Move the STHs topic along to encompass any unincorporated leaves.
+    offset := dbSTH.tree_size
+    batch := kafka.Read("Leaves", offset, batchSize)
+    for b := range batch {
+      db.Put("/<treeID>/leaves/<b.offset>", b.contents)
+    }
+
+    root := UpdateMerkleTreeAndBufferNodes(batch, treeRevision+1)
+    newSTH := STH{root, ...}
+    newSTH.treeRevision = dbSTH.treeRevision + 1
+    newSTH.expectedOffset = nextOffset
+    actualOffset := kafka.Append("STHs/<treeID>", newSTH)
+    if actualOffset != nextOffset {
+      glog.Warning("someone else wrote an STH while we were master")
+      tx.Abort()
+      return
+    }
+    newSTH.sthOffset = actualOffset
+    StoreSTHInDB(newSTH)
+    tx.Commit() // flush writes
+  } else {
+    // There is an STH one ahead of us that we're not caught up with yet.
+    for {
+      nextOffset++
+      nextSTH = kafka.Read("STHs/<treeID>", nextOffset)
+      if nextSTH.timestamp < ourSTH.timestamp || nextSTH.tree_size < ourSTH.tree_size {
+        glog.Errorf("should not happen - earlier STH with later offset")
+        return
+      }
+    }
+    // Read the leaves between what we have in our DB, and that STH...
+    leafRange := InclusiveExclusive(dbSTH.tree_size, nextSTH.tree_size)
+    batch := kafka.Read("Leaves", leafRange)
+    // ... and store it in our local DB
+    for b := range batch {
+      db.Put("<treeID>/leaves/<b.offset>", b.contents)
+    }
+    UpdateMerkleTreeAndStoreNodes(batch, treeRevision+1)
+    nextSTH.sthOffset = lastOffset
+    StoreSTHInDB(nextSTH)
+    tx.Commit() // flush writes
+    // We may still not be caught up, but that's for the next time around.
+  }
+}
+```
+
+##### Fit with storage interfaces
+
+LogStorage interfaces will need to be tweaked slightly, in particular:
+ - `UpdateSequencedLeaves` should be pulled out of `LeafDequeuer` and moved
+   into a `LeafSequencer` (or something) interface.
```

Comment:
Yup, more thinking req'd here, just haven't got to that yet.

It might be that a big Trillian deployment lever sets whether it's Trillian or the storage implementation which is handling the inter-cluster consistency, dunno yet... :)